subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskContainer.java

```diff
@@ -323,21 +364,47 @@ public Task create(String name, Closure configureClosure) {
         return registerTask(name, type, null, constructorArgs);
     }
 
-    private <T extends Task> TaskProvider<T> registerTask(final String name, final Class<T> type, @Nullable Action<? super T> configurationAction, Object... constructorArgs) {
+    private <T extends Task> TaskProvider<T> registerTask(final String name, final Class<T> type, @Nullable final Action<? super T> configurationAction, final Object... constructorArgs) {
         if (hasWithName(name)) {
             duplicateTask(name);
         }
-        DefaultTaskProvider<T> provider = Cast.uncheckedCast(getInstantiator().newInstance(TaskCreatingProvider.class, this, type, name, configurationAction, constructorArgs));
-        addLater(provider);
-        if (eagerlyCreateLazyTasks) {
-            provider.get();
-        }
-        return provider;
+        final InternalTaskInstanceId taskId = InternalTaskInstanceId.next();
+        return buildOperationExecutor.call(new CallableBuildOperation<TaskProvider<T>>() {
+            @Override
+            public BuildOperationDescriptor.Builder description() {
+                return new RegisterDetailsImpl(taskId, project, name, false).descriptor();
+            }
+
+            @Override
+            public TaskProvider<T> call(BuildOperationContext context) {
+                DefaultTaskProvider<T> provider = Cast.uncheckedCast(getInstantiator()
+                    .newInstance(TaskCreatingProvider.class, taskId, DefaultTaskContainer.this, type, name, configurationAction, constructorArgs)
+                );
+                addLater(provider);
+                if (eagerlyCreateLazyTasks) {
+                    provider.get();
```

Comment:
This should probably be outside the the register build operation. It might be better if we only fire a realize event and not a register + realize when eagerly creating lazy tasks (but not particularly important).

Reply:
We can rework it that way, but since the code went to some lengths to still create a provider etc I figured we'd want the ops to match.

Reply:
Unnested.